kernel/comps/network/src/lib.rs

```diff
@@ -148,33 +170,32 @@ pub fn all_devices() -> Vec<(String, NetworkDeviceRef)> {
 }
 
 static COMPONENT: Once<Component> = Once::new();
-pub(crate) static NETWORK_IRQ_HANDLERS: Once<
-    SpinLock<Vec<Arc<dyn NetDeviceIrqHandler>>, LocalIrqDisabled>,
-> = Once::new();
 
 #[init_component]
 fn init() -> Result<(), ComponentInitError> {
     let a = Component::init()?;
     COMPONENT.call_once(|| a);
-    NETWORK_IRQ_HANDLERS.call_once(|| SpinLock::new(Vec::new()));
+    SoftIrqLine::get(NETWORK_TX_SOFTIRQ_ID).enable(handle_tx_softirq);
+    SoftIrqLine::get(NETWORK_RX_SOFTIRQ_ID).enable(handle_rx_softirq);
     buffer::init();
     Ok(())
 }
 
-type NetDeviceIrqHandlerListRef =
-    Arc<SpinLock<Vec<Arc<dyn NetDeviceIrqHandler>>, LocalIrqDisabled>>;
-type NetworkDeviceRef = Arc<SpinLock<dyn AnyNetworkDevice, LocalIrqDisabled>>;
+type NetDevicePollCallbackListRef =
+    Arc<SpinLock<Vec<Arc<dyn NetDevicePollCallback>>, BottomHalfDisabled>>;
```

Comment:
Actually, I think it's a bit more desirable to retire the callback list altogether and let the kernel specify the softirq handler directly. The current implementation of `handle_{rx,tx}_softirq` in `comps/network` just invokes all devices' callbacks which is odd and useless.
 - The receive path is equivalent, as currently `comps/network` does nothing but execute the callback.
 - The send path needs some minor changes to perform `free_processed_tx_buffers` when the polling logic queries the device send status via `can_send`.

This can be a separate piece of work and done in a future PR if it's not appropriate to do it in this PR. But regardless, I'd like to know what you think about this? Do you think it is feasible?

Reply:
Certainly, I think it's reasonable.

The callback list is also redundant because it doesn't perform any tasks beyond polling the device. Similarly, Linux only polls the device in the softirq handler. It's enough for each device to register a poll function. Therefore, there's no need for us to maintain this list, as the issue can be addressed collectively.